File: backend/app/schemas/upload_preview.py
from datetime import date, datetime
from enum import Enum
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator
from pydantic.alias_generators import to_camel


class ApiModel(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True, extra="forbid")


class PreviewRangeMode(str, Enum):
    today = "today"
    yesterday = "yesterday"
    last_2_days = "last_2_days"
    custom = "custom"


class PreviewSource(str, Enum):
    plc = "plc"
    temperature = "temperature"


class PreviewProfile(str, Enum):
    default = "default"
    stage3_profile_a_bounded_full_scan = "stage3_profile_a_bounded_full_scan"
    large_source_operational = "large_source_operational"


STAGE3_PROFILE_A_BOUNDED_MAX_FILES = 3
STAGE3_PROFILE_A_BOUNDED_MAX_RUN_SECONDS = 300
STAGE3_PROFILE_A_BOUNDED_MAX_FILE_SECONDS = 120
LARGE_SOURCE_OPERATIONAL_MAX_FILES = 500
LARGE_SOURCE_OPERATIONAL_CHUNK_ROWS = 1000
LARGE_SOURCE_OPERATIONAL_MAX_RUN_SECONDS = 900
LARGE_SOURCE_OPERATIONAL_MAX_FILE_SECONDS = 300


class PreviewOptions(ApiModel):
    profile: PreviewProfile = PreviewProfile.default
    stable_lag_minutes: int = Field(default=3, ge=0, le=60)
    sample_rows: int = Field(default=200, ge=20, le=2000)
    chunk_rows: int = Field(default=20000, ge=1000, le=100000)
    max_files: int = Field(default=500, ge=1, le=5000)
    max_run_seconds: int = Field(default=120, ge=10, le=900)
    max_file_seconds: int = Field(default=30, ge=5, le=300)
    force_full_scan: bool = False

    @model_validator(mode="before")
    @classmethod
    def apply_profile_defaults(cls, data: Any) -> Any:
        if not isinstance(data, dict):
            return data
        profile = data.get("profile", PreviewProfile.default.value)
        profile_value = profile.value if isinstance(profile, PreviewProfile) else str(profile)
        if profile_value not in {
            PreviewProfile.stage3_profile_a_bounded_full_scan.value,
            PreviewProfile.large_source_operational.value,
        }:
            return data

        values = dict(data)

        def set_option(snake_name: str, camel_name: str, value: Any) -> None:
            if snake_name in values and camel_name not in values:
                values[snake_name] = value
            else:
                values[camel_name] = value

        if profile_value == PreviewProfile.stage3_profile_a_bounded_full_scan.value:
            set_option("force_full_scan", "forceFullScan", True)
            set_option("max_files", "maxFiles", STAGE3_PROFILE_A_BOUNDED_MAX_FILES)
            set_option("max_run_seconds", "maxRunSeconds", STAGE3_PROFILE_A_BOUNDED_MAX_RUN_SECONDS)
            set_option("max_file_seconds", "maxFileSeconds", STAGE3_PROFILE_A_BOUNDED_MAX_FILE_SECONDS)
        elif profile_value == PreviewProfile.large_source_operational.value:
            set_option("force_full_scan", "forceFullScan", False)
            set_option("max_files", "maxFiles", LARGE_SOURCE_OPERATIONAL_MAX_FILES)
            set_option("chunk_rows", "chunkRows", LARGE_SOURCE_OPERATIONAL_CHUNK_ROWS)
            set_option("max_run_seconds", "maxRunSeconds", LARGE_SOURCE_OPERATIONAL_MAX_RUN_SECONDS)
            set_option("max_file_seconds", "maxFileSeconds", LARGE_SOURCE_OPERATIONAL_MAX_FILE_SECONDS)
        return values


class PreviewCreateRequest(ApiModel):
    range_mode: PreviewRangeMode
    start_date: date | None = None
    end_date: date | None = None
    sources: list[PreviewSource] = Field(default_factory=lambda: [PreviewSource.plc])
    options: PreviewOptions = Field(default_factory=PreviewOptions)
    retry_of_run_id: str | None = None

    @model_validator(mode="after")
    def validate_date_range(self) -> "PreviewCreateRequest":
        if self.range_mode == PreviewRangeMode.custom:
            if self.start_date is None or self.end_date is None:
                raise ValueError("custom rangeMode requires startDate and endDate")
            if self.end_date < self.start_date:
                raise ValueError("endDate must be on or after startDate")
        return self


class PreviewItemStatus(str, Enum):
    target = "target"
    already_in_db = "already_in_db"
    partial_overlap = "partial_overlap"
    risky = "risky"
    excluded = "excluded"


class PreviewRunStatus(str, Enum):
    queued = "queued"
    running = "running"
    succeeded = "succeeded"
    partial_failed = "partial_failed"
    failed = "failed"
    cancelling = "cancelling"
    cancelled = "cancelled"
    timed_out = "timed_out"


class PreviewDbStatus(str, Enum):
    not_checked = "not_checked"
    reachable = "reachable"
    unreachable = "unreachable"
    query_failed = "query_failed"


class PreviewCreateResponse(ApiModel):
    preview_run_id: str
    status: PreviewRunStatus
    poll_url: str


class PreviewCancelResponse(ApiModel):
    preview_run_id: str
    status: PreviewRunStatus


class PreviewRunSummary(ApiModel):
    total: int = 0
    target: int = 0
    already_in_db: int = 0
    partial_overlap: int = 0
    risky: int = 0
    excluded: int = 0
    upload_rows: int = 0
    db_matched_rows: int = 0


class PreviewRunDto(ApiModel):
    preview_run_id: str
    status: PreviewRunStatus
    requested_at: datetime
    started_at: datetime | None = None
    finished_at: datetime | None = None
    db_status: PreviewDbStatus
    summary: PreviewRunSummary
    warnings: list[str] = Field(default_factory=list)
    requested_profile: PreviewProfile | None = None
    applied_profile: PreviewProfile | None = None
    auto_profile_reason: str | None = None
    timeout_stage: str | None = None
    timing: dict[str, Any] = Field(default_factory=dict)
    error_code: str | None = None
    error_message: str | None = None
    retry_of_run_id: str | None = None


class PreviewItemDto(ApiModel):
    preview_item_id: int
    status: PreviewItemStatus
    reason_code: str
    reason_text: str
    kind: PreviewSource
    folder_label: str
    filename: str
    path: str
    file_date: date | None = None
    size_bytes: int | None = None
    modified_at: datetime | None = None
    scan_mode: Literal["metadata", "sample", "full", "incomplete"]
    sample_row_count: int | None = None
    row_count: int | None = None
    local_key_count: int | None = None
    db_match_count: int | None = None
    upload_row_estimate: int | None = None
    first_timestamp: str | None = None
    last_timestamp: str | None = None
    device_ids: list[str] = Field(default_factory=list)
    issues: list[Any] = Field(default_factory=list)
    timeout_stage: str | None = None
    timing: dict[str, Any] = Field(default_factory=dict)
    error_code: str | None = None
    error_message: str | None = None


class PreviewPage(ApiModel):
    limit: int
    offset: int
    total_items: int


class PreviewDetailResponse(ApiModel):
    run: PreviewRunDto
    items: list[PreviewItemDto]
    page: PreviewPage


PreviewPageDto = PreviewPage
PreviewRunDetailResponse = PreviewDetailResponse

File: frontend/src/api/uploadPreview.ts
import { apiFetch } from "./client";

export type PreviewRangeMode = "today" | "yesterday" | "last_2_days" | "custom";
export type PreviewSource = "plc" | "temperature";
export type PreviewProfile =
  | "default"
  | "stage3_profile_a_bounded_full_scan"
  | "large_source_operational";
export type PreviewItemStatus =
  | "target"
  | "already_in_db"
  | "partial_overlap"
  | "risky"
  | "excluded";
export type PreviewRunStatus =
  | "queued"
  | "running"
  | "succeeded"
  | "partial_failed"
  | "failed"
  | "cancelling"
  | "cancelled"
  | "timed_out";
export type PreviewDbStatus = "not_checked" | "reachable" | "unreachable" | "query_failed";
export type PreviewSortKey = "status" | "fileDate" | "filename" | "uploadRows" | "modifiedAt";
export type SortOrder = "asc" | "desc";

export interface PreviewOptions {
  profile: PreviewProfile;
  stableLagMinutes: number;
  sampleRows: number;
  chunkRows: number;
  maxFiles: number;
  maxRunSeconds: number;
  maxFileSeconds: number;
  forceFullScan: boolean;
}

export interface PreviewCreateRequest {
  rangeMode: PreviewRangeMode;
  startDate: string | null;
  endDate: string | null;
  sources: PreviewSource[];
  options: PreviewOptions;
  retryOfRunId: string | null;
}

export interface PreviewCreateResponse {
  previewRunId: string;
  status: PreviewRunStatus;
  pollUrl: string;
}

export interface PreviewSummary {
  total: number;
  target: number;
  alreadyInDb: number;
  partialOverlap: number;
  risky: number;
  excluded: number;
  uploadRows: number;
  dbMatchedRows: number;
}

export interface PreviewRun {
  previewRunId: string;
  status: PreviewRunStatus;
  requestedAt: string;
  startedAt: string | null;
  finishedAt: string | null;
  dbStatus: PreviewDbStatus;
  summary: PreviewSummary;
  warnings: string[];
  requestedProfile?: PreviewProfile | null;
  appliedProfile?: PreviewProfile | null;
  autoProfileReason?: string | null;
  timeoutStage?: string | null;
  timing?: Record<string, unknown> | null;
  errorCode?: string | null;
  errorMessage?: string | null;
}

export interface PreviewItem {
  previewItemId: number;
  status: PreviewItemStatus;
  reasonCode: string;
  reasonText: string;
  kind: PreviewSource;
  folderLabel: string;
  filename: string;
  path: string;
  fileDate: string | null;
  sizeBytes: number | null;
  modifiedAt: string | null;
  scanMode: "metadata" | "sample" | "full" | "incomplete";
  rowCount: number | null;
  localKeyCount: number | null;
  dbMatchCount: number | null;
  uploadRowEstimate: number | null;
  firstTimestamp: string | null;
  lastTimestamp: string | null;
  deviceIds: string[];
  issues: string[];
  timeoutStage?: string | null;
  timing?: Record<string, unknown> | null;
  errorCode?: string | null;
  errorMessage?: string | null;
}

export interface PreviewPage {
  limit: number;
  offset: number;
  totalItems: number;
}

export interface PreviewResponse {
  run: PreviewRun;
  items: PreviewItem[];
  page: PreviewPage;
}

export interface PreviewQueryParams {
  status?: PreviewItemStatus | "all";
  q?: string;
  sort?: PreviewSortKey;
  order?: SortOrder;
  limit?: number;
  offset?: number;
}

export class ActivePreviewRunError extends Error {
  activePreviewRunId: string;

  constructor(activePreviewRunId: string) {
    super("A preview run is already active");
    this.name = "ActivePreviewRunError";
    this.activePreviewRunId = activePreviewRunId;
  }
}

const defaultOptions: PreviewOptions = {
  profile: "default",
  stableLagMinutes: 3,
  sampleRows: 200,
  chunkRows: 20_000,
  maxFiles: 500,
  maxRunSeconds: 120,
  maxFileSeconds: 30,
  forceFullScan: false,
};

export const stage3ProfileABoundedFullScanOptions: PreviewOptions = {
  ...defaultOptions,
  profile: "stage3_profile_a_bounded_full_scan",
  maxFiles: 3,
  maxRunSeconds: 300,
  maxFileSeconds: 120,
  forceFullScan: true,
};

export const largeSourceOperationalOptions: PreviewOptions = {
  ...defaultOptions,
  profile: "large_source_operational",
  chunkRows: 1_000,
  maxFiles: 500,
  maxRunSeconds: 900,
  maxFileSeconds: 300,
  forceFullScan: false,
};

export function createDefaultPreviewRequest(
  rangeMode: PreviewRangeMode,
  startDate: string | null,
  endDate: string | null,
): PreviewCreateRequest {
  return {
    rangeMode,
    startDate,
    endDate,
    sources: ["plc"],
    options: { ...defaultOptions },
    retryOfRunId: null,
  };
}

export function createLargeSourceOperationalPreviewRequest(
  rangeMode: PreviewRangeMode,
  startDate: string | null,
  endDate: string | null,
): PreviewCreateRequest {
  return {
    rangeMode,
    startDate,
    endDate,
    sources: ["plc"],
    options: { ...largeSourceOperationalOptions },
    retryOfRunId: null,
  };
}

export function createStage3ProfileABoundedFullScanPreviewRequest(
  rangeMode: PreviewRangeMode,
  startDate: string | null,
  endDate: string | null,
): PreviewCreateRequest {
  return {
    rangeMode,
    startDate,
    endDate,
    sources: ["plc"],
    options: { ...stage3ProfileABoundedFullScanOptions },
    retryOfRunId: null,
  };
}

function toCamelPreviewResponse(raw: any): PreviewResponse {
  const run = raw.run ?? {};
  const summary = run.summary ?? {};

  return {
    run: {
      previewRunId: run.previewRunId ?? run.preview_run_id,
      status: run.status,
      requestedAt: run.requestedAt ?? run.requested_at,
      startedAt: run.startedAt ?? run.started_at ?? null,
      finishedAt: run.finishedAt ?? run.finished_at ?? null,
      dbStatus: run.dbStatus ?? run.db_status,
      summary: {
        total: summary.total ?? 0,
        target: summary.target ?? summary.target_count ?? 0,
        alreadyInDb: summary.alreadyInDb ?? summary.already_in_db ?? summary.already_in_db_count ?? 0,
        partialOverlap:
          summary.partialOverlap ?? summary.partial_overlap ?? summary.partial_overlap_count ?? 0,
        risky: summary.risky ?? summary.risky_count ?? 0,
        excluded: summary.excluded ?? summary.excluded_count ?? 0,
        uploadRows: summary.uploadRows ?? summary.upload_rows ?? summary.upload_row_estimate ?? 0,
        dbMatchedRows: summary.dbMatchedRows ?? summary.db_matched_rows ?? summary.db_match_count ?? 0,
      },
      warnings: run.warnings ?? [],
      requestedProfile: run.requestedProfile ?? run.requested_profile ?? null,
      appliedProfile: run.appliedProfile ?? run.applied_profile ?? null,
      autoProfileReason: run.autoProfileReason ?? run.auto_profile_reason ?? null,
      timeoutStage: run.timeoutStage ?? run.timeout_stage ?? null,
      timing: run.timing ?? null,
      errorCode: run.errorCode ?? run.error_code ?? null,
      errorMessage: run.errorMessage ?? run.error_message ?? null,
    },
    items: (raw.items ?? []).map((item: any) => ({
      previewItemId: item.previewItemId ?? item.preview_item_id,
      status: item.status,
      reasonCode: item.reasonCode ?? item.reason_code,
      reasonText: item.reasonText ?? item.reason_text,
      kind: item.kind,
      folderLabel: item.folderLabel ?? item.folder_label,
      filename: item.filename,
      path: item.path,
      fileDate: item.fileDate ?? item.file_date ?? null,
      sizeBytes: item.sizeBytes ?? item.size_bytes ?? null,
      modifiedAt: item.modifiedAt ?? item.modified_at ?? null,
      scanMode: item.scanMode ?? item.scan_mode,
      rowCount: item.rowCount ?? item.row_count ?? null,
      localKeyCount: item.localKeyCount ?? item.local_key_count ?? null,
      dbMatchCount: item.dbMatchCount ?? item.db_match_count ?? null,
      uploadRowEstimate: item.uploadRowEstimate ?? item.upload_row_estimate ?? null,
      firstTimestamp: item.firstTimestamp ?? item.first_timestamp ?? null,
      lastTimestamp: item.lastTimestamp ?? item.last_timestamp ?? null,
      deviceIds: item.deviceIds ?? item.device_ids ?? [],
      issues: item.issues ?? [],
      timeoutStage: item.timeoutStage ?? item.timeout_stage ?? null,
      timing: item.timing ?? null,
      errorCode: item.errorCode ?? item.error_code ?? null,
      errorMessage: item.errorMessage ?? item.error_message ?? null,
    })),
    page: raw.page ?? { limit: 100, offset: 0, totalItems: raw.items?.length ?? 0 },
  };
}

function appendPreviewQuery(url: URL, params: PreviewQueryParams) {
  if (params.status && params.status !== "all") url.searchParams.set("status", params.status);
  if (params.q) url.searchParams.set("q", params.q);
  if (params.sort) url.searchParams.set("sort", params.sort);
  if (params.order) url.searchParams.set("order", params.order);
  if (params.limit) url.searchParams.set("limit", String(params.limit));
  if (params.offset) url.searchParams.set("offset", String(params.offset));
}

export async function createUploadPreview(
  request: PreviewCreateRequest,
): Promise<PreviewCreateResponse> {
  const response = await apiFetch(
    "/api/upload/preview",
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(request),
    },
    { mutating: true },
  );

  if (!response.ok) {
    if (response.status === 409) {
      const raw = await response.json().catch(() => null);
      const activePreviewRunId = raw?.detail?.activePreviewRunId;
      if (typeof activePreviewRunId === "string" && activePreviewRunId.length > 0) {
        throw new ActivePreviewRunError(activePreviewRunId);
      }
    }
    throw new Error("Upload preview could not be started");
  }

  const raw = await response.json();
  return {
    previewRunId: raw.previewRunId ?? raw.preview_run_id,
    status: raw.status,
    pollUrl: raw.pollUrl ?? raw.poll_url,
  };
}

export async function fetchUploadPreview(
  previewRunId: string,
  params: PreviewQueryParams,
): Promise<PreviewResponse> {
  const url = new URL(`/api/upload/preview/${previewRunId}`, window.location.origin);
  appendPreviewQuery(url, params);

  const response = await fetch(url.pathname + url.search);
  if (!response.ok) {
    throw new Error("Upload preview could not be loaded");
  }
  return toCamelPreviewResponse(await response.json());
}

export async function fetchLatestUploadPreview(
  params: PreviewQueryParams,
  completedOnly = false,
): Promise<PreviewResponse | null> {
  const url = new URL("/api/upload/preview/latest", window.location.origin);
  appendPreviewQuery(url, params);
  if (completedOnly) url.searchParams.set("completedOnly", "true");

  const response = await fetch(url.pathname + url.search);
  if (response.status === 404) return null;
  if (!response.ok) {
    throw new Error("Latest upload preview could not be loaded");
  }
  return toCamelPreviewResponse(await response.json());
}

export async function cancelUploadPreview(previewRunId: string): Promise<PreviewCreateResponse> {
  const response = await apiFetch(
    `/api/upload/preview/${previewRunId}/cancel`,
    { method: "POST" },
    { mutating: true },
  );
  if (!response.ok) {
    throw new Error("Upload preview could not be cancelled");
  }
  const raw = await response.json();
  return {
    previewRunId: raw.previewRunId ?? raw.preview_run_id,
    status: raw.status,
    pollUrl: raw.pollUrl ?? raw.poll_url ?? `/api/upload/preview/${previewRunId}`,
  };
}
